proposals/nnnn-sendable-function-values.md

```diff
@@ -0,0 +1,525 @@
+
+```swift
+extension StatefulTransformer {
+	// a nonisolated method trying to partially-apply an isolated method
+  nonisolated func outsideIsolationExample() -> [Data] {
+    let _: (Data) -> Data = self.transform
+    //                      ^~~~~~~~~~~~~~
+    // error: actor-isolated instance method 'transform' can not be partially applied
+	}
+}
+```
+
+Part of the need for this limitation is that knowledge of the isolation is effectively erased from the type signature when partially-applying the actor instance. That knowledge is required for non-async isolated methods, as their runtime representation cannot encode isolation enforcement alone. But, the limitation is not a fundamental barrier. It is feasible to create a `@Sendable` version of a partially-applied non-async actor method, by representing it with an `async` function value.
+
+Conceptually, a similar limitation comes up when performing a conversion that removes the `@MainActor` from a function type. Since it is generally unsafe to do so, the compiler emits a warning, such as in this example:
+
+```swift
+@MainActor func dropMainActor(_ f: @MainActor (Data) -> Data) -> [Data] {
+  return [4, 8, 15, 16, 23, 42].map(f)
+  // warning: converting function value of type '@MainActor (Data) -> Data' to '(Data) throws -> Data' loses global actor 'MainActor'
+}
+```
+
+But not all situations that drop the global-actor from a function value are unsafe. In the example above, the conversion that loses the global-actor happens while already on the same actor. By the same logic as our actor-method example, this should be safe if we do not allow that casted function to later leave the MainActor's isolation.
+
+These and other aspects of how Sendable and actor isolation interact with function values are the focus of this proposal.
+
+## Proposed solution
+
+This section provides a summary of the solutions and changes proposed for Swift. For complete details and caveats, see the Detailed design.
+
+### Inferring `@Sendable` for methods
+
+In [SE-302](0302-concurrent-value-and-concurrent-closures.md), the `@Sendable` attribute was introduced for both closures and named functions/methods. Beyond allowing the function value to cross actor boundaries, the attribute primarily influences the kinds of values that can be captured by the function. But methods of a nominal type cannot capture anything but the object instance itself. Furthermore, a non-local (i.e., a global or static) function cannot capture _anything_, because a reference to a global declaration is not considered a capture of that value. Thus, the proposed simplifications are:
+
+1. the inference of `@Sendable` on all methods of a type that conforms to `Sendable`.
+2. the inference of `@Sendable` on all non-local functions.
+3. the prohibition of marking a method `@Sendable` if the object type does not conform to `Sendable`.
```

Comment:
The indirect phrasing of these rules makes them a little difficult to understand. If I’m interpreting your text correctly, I’d suggest phrasing like this:

```suggestion
1. All methods of a type that conforms to `Sendable` shall be implicitly `@Sendable`.
2. All non-local functions shall be implicitly `@Sendable`.
3. It is not possible to mark a method `@Sendable` if the enclosing type does not conform to `Sendable`.
```

However, I’m still left with a few questions:

* Rule 1 only applies to *instance* methods, right?
* When Rule 2 talks about “non-local functions”, what exactly does it mean? I think it means free functions and static/class methods, but it’s not totally clear.
* Which functions do not fall under Rule 1, Rule 2, or Rule 3? That is, when would you still have a choice about whether a function is `@Sendable` and be able to explicitly mark it as such? (Maybe just local functions and perhaps closures?)